[enhance] improve dict in-predicate evaluate

englefly opened the pull request:
# Proposed changes
when column is dict encoded, in_predicate::evaluate() put dict code of IN_LIST into a set, and then compare the selected codes with column cells. The selected dict codes are put into a map, by which the evaluate time is O(logN)
It is better to use a std::vector<bool> to indicated if a dict word is in IN_LIST, by which the evaluate time shrinks to O(1).

test evn: ssb_flat 5G 
test sql: simplified from ssb q.3.3
SELECT count(c_city) FROM lineorder_flat WHERE C_CITY in ('UNITED KI1','UNITED KI5');

ShortPredEvalTime is decreased from 240ms to 170ms by average.

在测试环境中 ssb100-flat， 3BE q3.3 从291ms下降到248ms




Issue Number: close #xxx

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

englefly committed:
[enhance] improve dict in-predicate evaluate

englefly committed:
[enhance] improve dict in-predicate evaluate

yiguolei commented on be/src/olap/in_list_predicate.cpp:
```diff
@@ -161,12 +162,14 @@ IN_LIST_PRED_COLUMN_BLOCK_EVALUATE(NotInListPredicate, ==)
                 auto& data_array = dict_col.get_data();                                          \
-                auto dict_codes = dict_col.find_codes(_values);                                  \
+                std::vector<bool> selected;                                                  \
+                size_t dict_word_num = 0;                                                    \
```
dict_word_num is never used?

englefly replied:
I am not sure if there should be an assert(cell_value<dict_word_num).
I will add this assert.

englefly replied:
已经去掉了 dict_word_num 参数

HappenLee commented on be/src/olap/in_list_predicate.cpp:
```diff
@@ -132,13 +132,14 @@ IN_LIST_PRED_COLUMN_BLOCK_EVALUATE(NotInListPredicate, ==)
                         const auto& cell_value = data_array[idx];                                \
-                        bool ret = !null_bitmap[idx] &&                                          \
-                                   (dict_codes.find(cell_value) OP dict_codes.end());            \
+                        bool ret = !null_bitmap[idx] && (selected[cell_value] OP false);                    \
```
code format of `\`

englefly committed:
remove dict_word_num param from find_codes()

englefly committed:
remove dict_word_num param from find_codes()

yiguolei reviewed (dismissed):
LGTM

englefly committed:
add range check for dict type in-list-predicate

englefly committed:
add range check for dict type in-list-predicate

englefly committed:
add range check for dict type in IN_LIST_PREDICATE

englefly committed:
add range check for dict type in IN_LIST_PREDICATE

morningman committed:
[fix](fe-ut) Fix FE ut when enable vectorized engine (#9958)

Some node name in query explain result will be changes.
eg:
Aggregate -> VAggregate

xinyiZzz committed:
[fix][mem tracker] Fix logout load task mem tracker dcheck fail (#9943)

* fix tracker 0602

* fix format

zenoyang commented on be/src/vec/columns/column_dictionary.h:
```diff
@@ -258,9 +258,9 @@ class ColumnDictionary final : public COWHelper<IColumn, ColumnDictionary<T>> {
-    phmap::flat_hash_set<int32_t> find_codes(
-            const phmap::flat_hash_set<StringValue>& values) const {
-        return _dict.find_codes(values);
+    void find_codes(const phmap::flat_hash_set<StringValue>& values,
+                    std::vector<bool>& selected) const {
+        return _dict.find_codes(values, selected);
```
`find_codes` passes the result `selected` by reference, and not by `return`. This is not consistent with other interfaces, such as `find_code`, `find_code_by_bound`. I think it's better to keep it uniform.

BiteTheDDDDt committed:
[Feature] [Vectorized] Some pre-refactorings or interface additions for schema change (#9811)

Some pre-refactorings or interface additions for schema change

zhengshiJ committed:
[feature](statistics) Statistics derivation.Step 2:OtherNode implemen… (#9458)

closed #9644

Second step of statistics derivation: implementation of nodes other than scan_node.
The statistical information derivation interface of all nodes is uniformly placed in DeriveFactory.java.
Added one-sided to verify the derivation is correct.

Statistics derivation for each node is placed in its own *StatsDerive.java
detailed design: https://docs.google.com/document/d/1u1L6XhyzKShoyYRwFQ6kE1rnvY2iFwauwg289au5Qq0/edit

zy-kkk committed:
[revert](test) delete bitmap_index test case (#9999)

This test may cause endless loop

morningman committed:
[fix](coordinator) fix bug that unable to generate query profile (#10002)

This bug was introduced from #9720

Gabriel39 committed:
[Bug] runtime filter is not used as expected (#10001)

* [Bug] runtime filter is not used as expected

* update

yiguolei commented on be/src/olap/in_list_predicate.cpp:
```diff
@@ -161,12 +162,14 @@ IN_LIST_PRED_COLUMN_BLOCK_EVALUATE(NotInListPredicate, ==)
                     sel[new_size] = idx;                                                         \
                     const auto& cell_value = data_array[idx];                                    \
-                    auto result = (dict_codes.find(cell_value) OP dict_codes.end());             \
+                    assert(cell_value < selected.size());                                        \
```
Yes, doris uses dcheck instead of assert.

englefly replied:
assert only works in debug mode。 It will not affect the performance of release version

englefly committed:
use DCHECK instead of assert

englefly committed:
merge
